File: WC3/C++_homework/Unit/Unit.cpp
#include "Unit.h"

void Unit::ensureIsAlive() {
    if ( this->hitPoints < 1 ) {
        throw UnitIsDead();
    }
}
Unit::Unit(const std::string& name, int hp, int dmg) {
    if (dmg < 0 ) {
        this->damage = 0;
    } else {
        this->damage = dmg;
    }
    if ( hp < 1) {
        this->hitPoints = 1;

    } else {
        this->hitPoints = hp;
    }
    this->name = name;
    this->hitPointsLimit = this->hitPoints;
}
Unit::~Unit() {}

int Unit::getDamage() const {
    return this->damage;
}
int Unit::getHitPoints() const {
    return this->hitPoints;
}
int Unit::getHitPointsLimit() const {
    return this->hitPointsLimit;
}
const std::string& Unit::getName() const {
    return this->name;
}

void Unit::addHitPoints(int hp) {
    try { 
        this->ensureIsAlive();
    } catch ( UnitIsDead obj ) {
        std::cout << this->name <<" is dead. Cannot heal dead unit" << std::endl;
        return; 
    }
    if ( hp < 0 ) {
        std::cout << "added hp value must be positive" << std::endl;
        return;
    }
        if ( this->hitPoints + hp > this->hitPointsLimit ) {
            this->hitPoints = this->hitPointsLimit;
        } else {
            this->hitPoints += hp;
        }
}
void Unit::takeDamage(int dmg) {
    try { 
        this->ensureIsAlive();
    } catch ( UnitIsDead obj ) {
        std::cout << "Unit " << this->name << " is dead, cannot take damage." << std::endl;
        return; 
    }
    if ( dmg < 0 ) {
        std::cout << "Maybe you wanted to heal him instead?" << std::endl;
        return;
    }
    if ( dmg > this->hitPoints ) {
        this->hitPoints = 0;
        std::cout << "Unit " << this->name << " received " << dmg << " points of damage. Current hp = " << this->hitPoints << "/" << this->hitPointsLimit << ". Unit " << this->name << " is now dead." << std::endl;
    } else {
        this->hitPoints -= dmg;
        std::cout << "Unit " << this->name << " received " << dmg << " points of damage. Current hp = " << this->hitPoints << "/" << this->hitPointsLimit << std::endl; 
    }
}

void Unit::attack(Unit& enemy) {
    try {
        this->ensureIsAlive();
    } catch ( UnitIsDead obj ) {
        std::cout << "Attacker is dead, cannot attack" << std::endl;
        return;
    }
    try {
        enemy.ensureIsAlive();
    } catch ( UnitIsDead obj ) {
        std::cout << "Target is dead, cannot attack" <<std::endl;
        return;
    }
    std::cout << this->name << " tries to attack " << enemy.getName() << std::endl;

    enemy.takeDamage(this->damage);
    enemy.counterAttack(*this);

}

void Unit::counterAttack(Unit& enemy) {
    try {
        this->ensureIsAlive();
    } catch ( UnitIsDead obj ) {
        std::cout << "Counterattacker is dead, cannot attack" << std::endl;
        return;
    }
    std::cout << this->name << " is counterattacking " << enemy.getName() << std::endl;
    enemy.takeDamage(this->damage/2);
}



std::ostream& operator<<(std::ostream& out, const Unit& unit) {
    if ( unit.getHitPoints() > 0 ) {
    out << "Unit name is " << unit.getName() << std::endl << "Unit HP = " << unit.getHitPoints() << "/" << unit.getHitPointsLimit() << std::endl << "Unit damage = " << unit.getDamage();
    } else {
    out << "Unit name is " << unit.getName() << std::endl << "Unit is dead" << std::endl << "Unit damage = " << unit.getDamage();    
    }
    return out;
}

File: WC3/Draft_Homework with mains/Java_homework/Unit/Unit.java
public class Unit {
    private int damage;
    private int hitPoints;
    private int hitPointsMax;
    private String name;
    
    public void ensureIsAlive() throws UnitIsDeadException {
        if (this.hitPoints < 1) {
            throw new UnitIsDeadException("Unit " + this.name + " is dead.");
        }
    }
    
    public Unit(String name, int hp, int dmg) {
        if (hp < 1 ) {
            hp = 1;
        }
        if ( dmg < 0 ) {
            dmg = 0;
        }
        this.hitPointsMax = hp;
        this.hitPoints = hp;
        this.name = name;
        this.damage = dmg;
    }
    public Unit() {
        this.damage = 0;
        this.hitPoints = 1;
        this.hitPointsMax = 1;
        this.name = "Unknow unit";
    }
    
    public int getDamage() {
        return this.damage;
    }
    
    public int getHitPoints() {
        return this.hitPoints;
    }
    
    public int getHitPointsLimit() {
        return this.hitPointsMax;
    }
    public String getName() {
        return this.name;
    }
    
    public void addHitPoints(int hp) throws UnitIsDeadException {
        System.out.println("Healing phase");
        if (hp <= 0) {
            return;
        }
        try { 
            this.ensureIsAlive();

            if (this.hitPoints + hp > this.hitPointsMax) {
                System.out.println(this.name + " heals for " + (this.hitPointsMax - this.hitPoints) + " HP.");
                this.hitPoints = this.hitPointsMax;
            } else {
                this.hitPoints += hp;
                System.out.println(this.name + " heals for " + hp + " HP.");
            }
        } catch (UnitIsDeadException e) {
            System.out.println(e);
        }
    }
    
    public void takeDamage(int dmg) throws UnitIsDeadException {
        if (dmg < 0) {
            dmg = 0;
        }
        try { 
            this.ensureIsAlive();
            
            if (this.hitPoints < dmg) {
                this.hitPoints = 0;
            } else {
                this.hitPoints -= dmg;
            }
            
            if (this.hitPoints < 1) {
                System.out.println(this.name + " took " + dmg + " points of damage. He is dead now.");
            } else {
                System.out.println((this.name + " took " + dmg + " points of damage. HP: " + this.hitPoints + "/" + this.hitPointsMax));
            }
        } catch (UnitIsDeadException e) {
            System.out.println(e);
        }
    }
    
    public void attack(Unit enemy) throws UnitIsDeadException {
        System.out.println(this.name + " attack phase.");
        try {
            this.ensureIsAlive();
            enemy.ensureIsAlive();
            System.out.println(this.name + " is trying to attack " + enemy.name + " for " + this.damage + " damage.");
            enemy.takeDamage(this.damage);
        } catch (UnitIsDeadException e) {
            System.out.println(e);
        }
    }
    
    public void counterAttack(Unit enemy) throws UnitIsDeadException {
        try {
            System.out.println(this.name + " counterattack phase.");
            this.ensureIsAlive();
            enemy.ensureIsAlive();
            System.out.println(this.name + " is counterattacking " + enemy.name + " for " + this.damage/2 + " damage.");
            enemy.takeDamage(this.damage/2);
        } catch (UnitIsDeadException e) {
            System.out.println(e);
        }

    }
    
    public String toString() {
        return "Unit name: " + this.name + "\n" + "HP: " + this.hitPoints + "/" + this.hitPointsMax + "\n" + "Damage: " + this.damage + "\n";
    }
    
}